File: src/example/c/file.c
#include <stdio.h>
#include "modserver.h"

int run(servlet *s)
{
  FILE *f = fopen("example/c/file.c", "r");
  if (f)
  {
    fseek(f, 0, SEEK_END);
    long length = ftell(f);
    rewind(f);
    char str_length[128];
    snprintf(str_length, sizeof(str_length), "%lu", length);
    set_header(s, "Content-Length", str_length);
    set_header(s, "Content-Type", "text/plain; charset=UTF-8");
    char buf[BUFSIZ];
    while (1)
    {
      size_t bytes_read = fread(buf, 1, sizeof(buf), f);
      if (bytes_read > 0)
      {
        rwrite(s, buf, bytes_read);
      }
      else
      {
        break;
      }
    }
    fclose(f);
  }
  else
  {
    rprintf(s, "file not found");
  }
  return 0;
}

File: src/example/lua/file.lua
local servlet = {}

function servlet:run()
  local f = io.open("example/lua/file.lua", "r")
  if f then
    local length = f:seek("end")
    f:seek("set")
    self:set_content_length(length)
    self:set_header("Content-Type", "text/plain; charset=UTF-8");
    while true do
      local buf = f:read(4096)
      if buf then
        self:rwrite(buf)
      else
        break
      end
    end
    f:close()
  else
    self:rwrite("file not found")
  end
end

return servlet
